designs/0034-jacobian.md

```diff
@@ -0,0 +1,381 @@
+ real<constraint upper_bound(u_bound)> c = a;
+}
+
+
+parameters {
+ // User-defined constraint applied to parameters alpha and b
+ real<constraint upper_bound(u_bound)> alpha;
+ vector<constraint upper_bound(u_bound)>[N] b;
+ // In these cases this is the same as calling
+ // real<upper=u_bound> alpha;
+ // vector<upper=u_bound>[N] b;
+}
+```
+
+In this example, `upper_bound` is a user-defined constraint that is applied to the data and parameters.
+The `validate()` function from `upper_bound` is used to validate `a_validated` in the `transformed data` block, and the transformations defined in `upper_bound` are applied to `alpha` and `b` in the parameters block.
+
+### Error Messages
+
+Inside of a constraint, if a `constrain` function is defined then an `unconstrain` function must also be defined with the same signature as `constrain`.
+
+```
+constraint(vector x, real upper_bound) for {CONSTRAINT_NAME} does not have an associated `unconstrain(vector x, real upper_bound)`. If you define one then you must define both!
+```
+
+If a user defined constraint type is used in `data`, `transformed data`, or `transformed parameters`, or `generated quantities` then the constraint must have a validate function for that given type. 
+For example, if a user calls a constraint named `lower_bound` tha does not have a `validate()` function then the error message will look like:
+
+```
+// vector<constrain lower_bound(lb)> x;
+a `lower_bound` constraint was given to `x`, but `lower_bound` does not have a validate(vector, real) function.
+```
+
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+The first thing we would do is add the `jacobian` keyword to the Stan language. 
+Like `target`, `jacobian` would only be available on the left hand side of statements to be accumulated into. 
+Its domain is restricted to the `transformed parameters` and `constraints` block.
+The main difference between `jacobian` and `target` is that the jacobian accumulation is optional for Stan algorithms. If an algorithm wants to do something like maximum likelihood estimation then it is possible to turn off the Jacobian accumulation.
+
+This section will break down the model given in the summary section to show what the Stan compiler requires piece by piece. 
+First, let's look at the `constraints` block where we define our constrain and unconstrain functions
+
+## Constraint Block
+
+```stan
+constraints {
+ upper_bound {
+   constrain(real x, real upper_bound) {
+     jacobian += x;
+     return upper_bound - exp(x);
+   }
+   real unconstrain(real x, real upper_bound) {
+     return log(upper_bound - x);
+   }
+   // Used for data
+   int validate(real x, real upper_bound) {
+     // For most this will be
+     return x < upper_bound;
+   }
+ }
+}
+```
+
+The new `constraints` block will need to be added to the compiler's parser, validator, Middle Implementation Layer (MIR), and c++ MIR. 
```

Comment:
I don't see why this would persist down to the C++ IR level. At that point it would just be functions with template arguments for the jacobian, right?

Even having it in the MIR seems optional, honestly. This could all be desugared at the AST->MIR level (assuming that either jacobian+= works in `_lp` functions or we added another suffix)